pmd-core/src/main/java/net/sourceforge/pmd/properties/PropertyBuilder.java

```diff
@@ -0,0 +1,377 @@
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.StringUtils;
+
+import net.sourceforge.pmd.properties.PropertyBuilder.GenericCollectionPropertyBuilder.Supplier;
+import net.sourceforge.pmd.properties.builders.PropertyDescriptorBuilder;
+import net.sourceforge.pmd.properties.constraints.PropertyConstraint;
+
+
+/**
+ * Base class for generic property builders.
+ * Note: from 7.0.0 on, all property builders will
+ * extend this class instead of {@link PropertyDescriptorBuilder}.
+ *
+ * @param <B> Concrete type of this builder instance
+ * @param <T> Type of values the property handles
+ *
+ * @author Clément Fournier
+ * @since 6.10.0
+ */
+public abstract class PropertyBuilder<B extends PropertyBuilder<B, T>, T> {
+
+    private static final Pattern NAME_PATTERN = Pattern.compile("[a-zA-Z][\\w-]*");
+    private final Set<PropertyConstraint<? super T>> validators = new LinkedHashSet<>();
+    protected boolean isDefinedExternally;
+    private String name;
+    private String description;
+    private float uiOrder = 0f;
+    private T defaultValue;
+
+
+    PropertyBuilder(String name) {
+
+        if (StringUtils.isBlank(name)) {
+            throw new IllegalArgumentException("Name must be provided");
+        } else if (!NAME_PATTERN.matcher(name).matches()) {
+            throw new IllegalArgumentException("Invalid name '" + name + "'");
+        }
+        this.name = name;
+    }
+
+
+    // will maybe be scrapped
+    @Deprecated
+    void setDefinedExternally(boolean bool) {
+        this.isDefinedExternally = bool;
+    }
+
+
+    Set<PropertyConstraint<? super T>> getConstraints() {
+        return validators;
+    }
+
+
+    String getDescription() {
+        return description;
+    }
+
+
+    @Deprecated
+    float getUiOrder() {
+        return uiOrder;
+    }
+
+
+    T getDefaultValue() {
+        return defaultValue;
+    }
+
+
+    /**
+     * Specify the description of the property.
+     *
+     * @param desc The description
+     *
+     * @return The same builder
+     */
+    @SuppressWarnings("unchecked")
+    public B desc(String desc) {
+        if (StringUtils.isBlank(desc)) {
+            throw new IllegalArgumentException("Description must be provided");
+        }
+        this.description = desc;
+        return (B) this;
+    }
+
+
+    /**
+     * Add a constraint on the values that this property may take.
+     * The validity of values will be checked when parsing the XML,
+     * and invalid values will be reported. A rule will never be run
+     * if some of its properties violate some constraints.
+     *
+     * @param constraint The constraint
+     *
+     * @return The same builder
+     */
+    // TODO we could probably specify the order of execution of constraints come 7.0.0, for now this remains unspecified
```

Comment:
Indeed it wouldn't matter if constraints are truly independent -> in which case we should specify that. I think it makes sense, and it would allow to report all errors on a value, like you suggested for rules.